Add shadow back for blurred/transparent window on macOS

alphaArgon opened the pull request:
Closes #15383
Closes #10993

`NSVisualEffectView` is an official API for implementing blur effects and, by traversing the layers, we **can remove the background color** that comes with the view. This avoids using private APIs and aligns better with macOS’s native design.

Currently, `GPUIView` serves as the content view of the window. To add the blurred view, `GPUIView` is downgraded to a subview of the content view, placed at the same level as the blurred view.

Release Notes:

- Fixed the missing shadow for blurred-background windows on macOS.

alphaArgon committed:
Add shadow back for blurred window on macOS

Wrap GPUIView inside a content view. Add a custom subclass of `NSVisualEffectView` to implement the blur effects.

alphaArgon commented:
@cla-bot check

notpeter committed:
Fix formatting

jansol commented:
Does this actually work across macOS versions? I remember coming across this approach when investigating blurred windows initially and it was mentioned that it only worked on specific versions of macOS (well, Cocoa).

jansol commented on crates/gpui/src/platform/mac/window.rs:
```diff
@@ -1007,8 +1006,27 @@ impl PlatformWindow for MacWindow {
                 NSColor::clearColor(nil)
             };
             this.native_window.setBackgroundColor_(clear_color);
-            let window_number = this.native_window.windowNumber();
```
The `this.native_window.setHasShadow_(opaque);` above should also be removed I think?

alphaArgon commented:
@jansol 

> Does this actually work across macOS versions

I’m sorry for being late with the full-version testing, and I found my implementation doesn’t work on macOS Catalina. Prior to macOS Big Sur, `NSVisualEffectView` doesn’t own concrete sublayers; instead, it transfers all the rendering work to a `CAProxyLayer`.

> I remember coming across this approach when investigating blurred windows initially and it was mentioned that it only worked on specific versions of macOS (well, Cocoa) because the layers/names changed.

As long as the proxy layer is not used, checking `filters` by name is robust: unlike layers, filter types (`CAFilter`, `CIFilter`) use `name` to identify what effects they should do, which is a relatively stable property.

Another thing is setting the window background color to a fully transparent color causes the shadow being drawn differently. A bitmap of the content is taken, made monotone and blurred, and then attached as the shadow. If the opacity of the content changes, the shadow won’t get updated unless an `-[NSWindow invalidateShadow]` is called. Therefore, for clear background colors we must disable the shadow to avoid visual inconsistency.

However, as long as the window is `titled` and the background color is partially opaque, say having an alpha of 0.01, the shadow behaves as if it’s in normal windows. Maybe I made things complicated that all we need is a very-low-alpha color?

jansol commented:
> As long as the proxy layer is not used, checking filters by name is robust: unlike layers, filter types (CAFilter, CIFilter) use name to identify what effects they should do, which is a relatively stable property.

Thanks for the clarification. Can you add this explanation as a comment in the code as well? As for macOS support... Zed *does* only support some number of versions going back but I'm not sure what the official policy for the cutoff is. Maybe a good approach would be to fall back to the current implementation for versions older than Big Sur?


> Another thing is setting the window background color to a fully transparent color causes the shadow being drawn differently. A bitmap of the content is taken, made monotone and blurred, and then attached as the shadow. If the opacity of the content changes, the shadow won’t get updated unless an -[NSWindow invalidateShadow] is called. Therefore, for clear background colors we must disable the shadow to avoid visual inconsistency.

Hmm.. I do remember getting "phantom text" even with a background alpha of like 0.5. But that was on Catalina, maybe they changed the shadow rendering logic after that.

> However, as long as the window is `titled` and the background color is partially opaque, say having an alpha of 0.01, the shadow behaves as if it’s in normal windows. Maybe I made things complicated that all we need is a very-low-alpha color?

I'm hesitant to limit background opacity, some people did deliberately set it to exactly zero for some specific scenarios. But disabling shadows when background alpha is zero would probably be an acceptable compromise. Still an improvement over status quo and if someone *really* wants a fully transparent background the shadow would look awkward anyway. But that is ultimately something the designer types of the Zed team have to decide.

alphaArgon commented:
@jansol

> Hmm.. I do remember getting "phantom text" even with a background alpha of like 0.5. But that was on Catalina, maybe they changed the shadow rendering logic after that.

Friends and I tested setting the background to `-[NSColor colorWithWhite:0 alpha:0.0001]`. This color is invisible to the human eye while not technically being `clear`, and here’s† how it works on macOS High Sierra and Ventura. The shadow works fine and breaks not on Mission Control. Since `GPUIWindow` is a standard `titled` window, “phantom text” won’t appear with a non-zero-alpha background color.

> I'm hesitant to limit background opacity, some people did deliberately set it to exactly zero

There are two kinds of background colors: the settings’ `background` (rendered by `GPUIView`), and the background color of `CGUIWindow`; they are independent. Currently the `CGUIWindow`’s background color is set to `-[NSColor clearColors]` when the settings’ `background.appearance` is not `opaque` — that’s the problem. This also makes the backdrop blur disappear if the setting’s `background` is set to to `#xxxxxx00`, which I don’t think is by design (no logic found in the code) and appears to be a bug of the Quartz Compositor.

> but I'm not sure what the official policy for the cutoff is.

Currently supports macOS Catalina. Sadly. Using two implementations could be hard to maintain. If we want the shadow back, replacing `NSColor::clearColor(nil)` to somecolor like `NSColor::colorWithSRGBRed_green_blue_alpha_(nil, 0f64, 0f64, 0f64, 0.0001)` could be the best approach. I’m opening an issue about *whether* and *where* the shadow should be added back.

---

† The screenshots:

![image](https://github.com/user-attachments/assets/33319940-9654-42e2-9dcf-7bb9089f40d2)

alphaArgon committed:
Merge branch 'main' of https://github.com/alphaArgon/zed

alphaArgon commented:
> Mmm okay, maybe a black background with 0.0001 alpha is the way to go then. I don't have access to any macs these days so I can't check myself. Thanks for doing the in depth research though! 🙏

Well it seems that we still need to use an `NSVisualEffectView`... The maximum *effective* radius for `CGSSetWindowBackgroundBlurRadius` is about 15; setting it to the current value (80) causes a clamp. It’s less blurry than the `NSVisualEffectView` (30):

![image](https://github.com/user-attachments/assets/d97398d8-698f-4e61-be8a-688f316c09b4)

alphaArgon committed:
Refine macOS window background blurring and shadowing.

- Use a hybrid way to implement the blurred background based on the system version
- Fix blurring effect missing when user uses a clear background (bug by WindowServer)

alphaArgon committed:
Merge branch 'main' into main

alphaArgon committed:
Format fix

jansol reviewed (commented):
I no longer have access to a mac for testing but by and large this looks okay. I'm not a huge fan of relying on long deprecated materials, I'd rather find the tint effect and pluck it out like the saturation layer, but I have no idea how that is implemented internally.

The fallback to the old solution for old macOS versions is nice but did you make sure it's not running into the issue where performance of resizing and mission control would be very sluggish due to shadows getting recalculated too frequently?

alphaArgon committed:
Remove use of deprecated materials

- Remove desktop tinting by hiding the `CAChameleonLayer` instead of using a deprecated material.
- Fix architecture incompatibility of ObjC BOOL.
- Restore the legacy blur radius

notpeter committed:
Merge branch 'main' into alphaArgon/main

rjeli commented:
I can confirm this fixes it on 15.5

![image](https://github.com/user-attachments/assets/c6a4302f-f300-4a70-8ee8-933a7a773a58)

rjeli commented:
i can confirm fixed on sequoia 15.5!
![image](https://github.com/user-attachments/assets/0699486f-530b-428e-9e17-719764653921)

SomeoneToIgnore reviewed (commented):
I have tried to test this but have not succeeded due to 0 knowledge on how to test it.

Thank you for coming up and posting a screenshot of the rendering with all the related information.

For my amateur view, seems to work as expexted.
I'd love to get a confirmation from @jansol and/or the PR creator; and I"d really appreciate a short paragraph in the https://github.com/zed-industries/zed/blob/main/docs/src/themes.md before merging.

jansol commented:
> I have tried to test this but have not succeeded due to 0 knowledge on how to test it.

The easiest way is to check the corner case(s):
1. install the Glazier extension
2. switch to the Moss Fett theme
3. add `"background": "#00000000"` to `theme_overrides`
4. toggle macOS back and forth between light and dark mode

If everything works perfectly, there should be no tinting/lightening/darkening in the window background with the exception of the shadow behind the window, and more importantly it should not change when switching between light and dark mode. The latter should be easy to verify with a screenshot or color picker tool.

I'm not entirely sure what you'd expect from that paragraph in the docs? Assuming everything works as described here, the only takeaway should really be that transparent windows will still always have a shadow affecting what is seen through the window.

rjeli commented:
Yeah this just fixes [this bug](https://github.com/harmtemolder/Solarized.zed/issues/2#issue-3183807256) for me, here's `main` for comparison:
![image](https://github.com/user-attachments/assets/7512470e-b127-4748-b88e-a14bfea522f2)

> The easiest way is to check the corner case(s):

![image](https://github.com/user-attachments/assets/2a48dc73-7167-4cf6-9c32-892923b53a06)
![image](https://github.com/user-attachments/assets/e49df63a-9f20-4bea-9563-c1e77bf1ec62)
https://github.com/user-attachments/assets/42614795-2f0b-4866-8b4a-38b6748282f9

lgtm

jansol commented:
Perfect!

Can you also post a comparison with `"background.appearance": "transparent"`? The blurred background looks like it ignores the shadow behind the window. (not relevant to this PR but I'd appreciate a comparison point for #21341)

rjeli commented:
yep:

![image](https://github.com/user-attachments/assets/51f5e187-4674-4b6e-9077-212d9492ebee)
![image](https://github.com/user-attachments/assets/40540542-5609-4891-a768-7075acc8ff20)

video: https://github.com/user-attachments/assets/f5d43723-1a0e-45af-ac96-6927ba53df73

SomeoneToIgnore reviewed (approved):
Thank you so much for testing and explaining.

> I'm not entirely sure what you'd expect from that paragraph in the docs? 

What I hoped for is a "how to make Zed transparent" bullet, maybe coupled with more configurations that allow to adjust Zed's appearance in a non-default way.
